File: 2021/2021_23.py
import copy
import math

class burrow:
    # Map is hardcoded
    burrowsPos = (2, 4, 6, 8)
    anthropodStr = ("A", "B", "C", "D")

    def __init__(self, hallway, burrows):
        self.hallway = hallway
        self.burrows = burrows
        self.energy_usage = 0

    def __str__(self):
        result = ""
        for h in self.hallway:
            if type(h) == int:
                result += self.anthropodStr[h]
            else:
                result += h
        result += "\n "
        for b in range(len(self.burrows)):
            if len(self.burrows[b]) == 2:
                result += " " + self.anthropodStr[self.burrows[b][1]]
            else:
                result += "  "
        result += "  \n "
        for b in range(len(self.burrows)):
            if len(self.burrows[b]) > 0:
                result += " " + self.anthropodStr[self.burrows[b][0]]
            else:
                result += "  "
        result += "  \n"
        return result

    def __eq__(self, other):
        return self.hallway == other.hallway and self.burrows == other.burrows and self.energy_usage == other.energy_usage

    def __ne__(self, other):
        return not self.__eq__(other)

    # def __hash__(self):
    #     return 

    def __copy__(self):
        B = burrow(list(self.hallway), copy.deepcopy(self.burrows))
        B.energy_usage = self.energy_usage
        return B

    def isDone(self):
        for a in range(4):
            if self.burrows[a] != [a, a]:
                return False
        return True

    def score(self):
        result = -self.energy_usage
        for b in range(len(self.burrows)):
            B = self.burrows[b]
            if len(B) > 0:
                if B[0] == b: result += 2000
            if len(B) > 1:
                if B[1] == b: result += 2000
        return result

    def moveToBurrow(self, pos): # pos in hallway
        if self.hallway[pos] == '.':
            return None # Not an anthropod
        a = self.hallway[pos] # Anthropod

        if self.burrows[a] == []:
            steps = 2
        elif self.burrows[a] == [a]:
            steps = 1
        else:
            return None # No room or other anthropods are still in there
        
        # Check path (path inside burrow is clear)
        inc = 1 if self.burrowsPos[a] > pos else -1
        p = pos
        while p != self.burrowsPos[a]:
            p += inc
            steps += 1
            if self.hallway[p] != '.':
                return None # Blocked
        
        # Done! We can make that move
        result = copy.copy(self)
        result.hallway[pos] = '.'
        result.burrows[a].append(a)
        assert(len(result.burrows[a]) <= 2)
        result.energy_usage += steps * pow(10, a)
        return result

    def moveToHallway(self, b): # b = burrow index
        if self.burrows[b] == []:
            return None # Empty burrow
        if self.burrows[b] == [b]:
            return None # or only filled with the correct anthropods
        if self.burrows[b] == [b, b]:
            return None
        a = self.burrows[b][-1]
        

        # Now find all possibilities where it can go
        result = []

        p = self.burrowsPos[b]
        # Move into hallway
        steps = (3 - len(self.burrows[b]))
        while p > 0:
            p -= 1
            steps += 1
            if self.hallway[p] != '.':
                break # Blocked
            if p in self.burrowsPos:
                continue # Move beyond burrow exits
            new_state = copy.copy(self)
            new_state.hallway[p] = a
            new_state.burrows[b].pop() # pops last item
            new_state.energy_usage += steps * pow(10, a)
            result.append(new_state)

        p = self.burrowsPos[b]
        # Move into hallway
        steps = (3 - len(self.burrows[b]))
        while p < len(self.hallway) - 1:
            p += 1
            steps += 1
            if self.hallway[p] != '.':
                break # Blocked
            if p in self.burrowsPos:
                continue # Move beyond burrow exits
            new_state = copy.copy(self)
            new_state.hallway[p] = a
            new_state.burrows[b].pop() # pops last item
            new_state.energy_usage += steps * pow(10, a)
            result.append(new_state)
        return result

    def possibilities(self):
        result = []
        # All possibilities where an amphipod moves out of a burrow
        for b in range(len(self.burrowsPos)):
            new_states = self.moveToHallway(b)
            if new_states:
                result += new_states
        
        # All possibilities where an amphipod moves into its own burrow
        for pos in range(len(self.hallway)):
            new_state = self.moveToBurrow(pos)
            if new_state:
                result.append(new_state)

        return result

# Testcase. Costs 12521 energy
#############
#...........#
###B#C#B#D###
  #A#D#C#A#
  #########
Btest0 = burrow(['.','.','.','.','.','.','.','.','.','.','.'], [[0, 1], [3, 2], [2, 1], [0, 3]])
Btest0_next = Btest0.moveToHallway(2)
# for B in Btest0_next:
#     print(B)
Btest1 = Btest0_next[1]
assert(Btest1.energy_usage == 40)
assert(Btest1.moveToHallway(2) == None)
Btest1_next = Btest1.moveToHallway(1)
# for B in Btest1_next:
#     print(B)
Btest2 = Btest1_next[0].moveToBurrow(5)
assert(Btest2.energy_usage == 440)
assert(Btest2.moveToHallway(2) == None)

#############  A = 0
#...........#  B = 1
###C#A#B#D###  C = 2
  #C#A#D#B#    D = 3
  #########
B = burrow(['.','.','.','.','.','.','.','.','.','.','.'], [[2, 2], [0, 0], [3, 1], [1, 3]])

# Search for the best answer
frontier = [B]
visited = []
lowest_energy = math.inf
while frontier:
    frontier.sort(key=lambda B: B.score(), reverse=False)
    print("lowest energy: " + str(lowest_energy) + " todo: " + str(len(frontier)) + " checked: " + str(len(visited)))
    # print(frontier[-1])
    next_burrow = frontier.pop() # Best score at the back

    # Consider the best one
    is_new = True
    found_better = False
    for n in range(len(visited)):
        if visited[n].hallway == next_burrow.hallway and visited[n].burrows == next_burrow.burrows:
            is_new = False
            if visited[n].energy_usage > next_burrow.energy_usage:
                visited[n] = next_burrow
                found_better = True
            break
    if is_new:
        visited.append(next_burrow)
    elif not found_better:
        continue
    
    # Propagate into new states
    next_states = next_burrow.possibilities()
    for S in next_states:
        if S.energy_usage >= lowest_energy:
            continue # Skip this one, can't be better
        elif S.isDone():
            if S.energy_usage < lowest_energy:
                lowest_energy = S.energy_usage
                print("Lowest energy so far: " + str(lowest_energy) + " checked: " + str(len(visited)))
        else:
            frontier.append(S)

# answer > 6736. != 12766

File: 2021/2021_23.cpp
#include <algorithm>
#include <array>
#include <cmath>
#include <iostream>
#include <optional>
#include <unordered_set>
#include <string>

constexpr auto kNrBurrows = 4;
constexpr auto kNumToChar = "0123";
constexpr std::array<int, kNrBurrows> kBurrowsPos = {2, 4, 6, 8};

constexpr int ctoi(char c) {
    return static_cast<int>(c) - '0';
}

constexpr auto kAmphNumToName = "ABCD";
constexpr char AmphName(char aa) {
  return kAmphNumToName[ctoi(aa)];
}

template <int tBurrowDepth>
class Burrow {
  public:
    Burrow(std::string hallway, std::array<std::string, kNrBurrows> burrows)
      : hallway_{std::move(hallway)}
      , burrows_{std::move(burrows)}
      , energy_usage_{0} {
    }

    bool IsDone() const {
        for (int n = 0; n < kNrBurrows; ++n)
            if (std::count(burrows_[n].begin(), burrows_[n].end(), kNumToChar[n]) != tBurrowDepth) {
                return false;
            }
        return true;
    }

    bool Equivalent(Burrow const& other) const {
      return hallway_ == other.hallway_ && burrows_ == other.burrows_;
    }

    long long Score() const {
      long long result = 0;
      for (int b = 0; b < kNrBurrows; ++b) {
        for (char bb : burrows_[b]) {
          if (bb == kNumToChar[b]) {
            result += 10000;
          } else {
            result -= 5000;
          }
        }
      }
      return result;
    }

    std::optional<Burrow> MoveToBurrow(int pos) const {
      if (hallway_[pos] == '.') {
          return {};
      }

      int a = ctoi(hallway_[pos]);
      if (burrows_[a].size() == tBurrowDepth) {
          return {}; // No room
      }

      char aa = kNumToChar[a];
      if (std::any_of(burrows_[a].begin(), burrows_[a].end(), [aa](char c) { return c != aa; })) {
        return {}; // Wrong anthropods in there
      }

      // Check path
      int steps = 1 + tBurrowDepth - burrows_[a].size(); // One initial step, plus depth into burrow
      int inc = kBurrowsPos[a] > pos ? 1 : -1;
      for (int p = pos + inc; p != kBurrowsPos[a]; p += inc) {
          ++steps;
          if (hallway_[p] != '.') {
              return {}; // Blocked
          }
      }

      // Done! We can make that move.
      auto new_burrow = *this;
      new_burrow.hallway_[pos] = '.';
      new_burrow.burrows_[a].push_back(aa);
      new_burrow.energy_usage_ += steps * std::pow(10, a);
      return new_burrow;
    }

    std::vector<Burrow> MoveToHallway(int b) const {
      // b = burrow index
      if (burrows_[b].empty()) {
          return {};
      }

      char bb = kNumToChar[b];
      if (std::all_of(burrows_[b].begin(), burrows_[b].end(), [bb](char c){ return c == bb; })) {
          return {}; // Only anthropods in there that should stay there
      }

      char aa = burrows_[b].back();
      int a = ctoi(aa);
      std::vector<Burrow> result;

      int steps = 1 + tBurrowDepth - burrows_[b].size();
      for (int p = kBurrowsPos[b] + 1; p < static_cast<int>(hallway_.size()); ++p) {
          ++steps;
          if (hallway_[p] != '.') {
              break; // Blocked
          }
          if (std::count(kBurrowsPos.begin(), kBurrowsPos.end(), p) > 0) {
              continue; // Move beyond burrow exits
          }

          auto new_burrow = *this;
          new_burrow.hallway_[p] = aa;
          new_burrow.burrows_[b].pop_back();
          new_burrow.energy_usage_ += steps * std::pow(10, a);
          result.push_back(new_burrow);
      }

      steps = 1 + tBurrowDepth - burrows_[b].size();
      for (int p = kBurrowsPos[b] - 1; p >= 0; --p) {
          ++steps;
          if (hallway_[p] != '.') {
              break; // Blocked
          }
          if (std::count(kBurrowsPos.begin(), kBurrowsPos.end(), p) > 0) {
              continue; // Move beyond burrow exits
          }

          auto new_burrow = *this;
          new_burrow.hallway_[p] = aa;
          new_burrow.burrows_[b].pop_back();
          new_burrow.energy_usage_ += steps * std::pow(10, a);
          result.push_back(new_burrow);
      }

      return result;
    }

    std::vector<Burrow> Possibilities() const {
      std::vector<Burrow> result;
      for (int b = 0; b < kNrBurrows; ++b) {
          auto options = MoveToHallway(b);
          for (const auto& option : options) {
            result.push_back(option);  
          }
      }
      for (int p = 0; p < static_cast<int>(hallway_.size()); ++p) {
        auto option = MoveToBurrow(p);
        if (option) {
            result.push_back(*option);
        }
      }
      return result;
    }

    std::string hallway_;
    std::array<std::string, kNrBurrows> burrows_;
    long long energy_usage_;
};

template <int T>
std::ostream& operator<<(std::ostream& o, Burrow<T> const& B) {
  for (auto c : B.hallway_) {
    if (c == '.')
      o << c;
    else
      o << AmphName(c);
  }
  o << "\n";
  for (int n = T - 1; n >= 0; --n) {
    o << "  ";
    for (int b = 0; b < kNrBurrows; ++b) {
      if (static_cast<int>(B.burrows_[b].size()) > n) {
        o << AmphName(B.burrows_[b][n]) << " ";
      }
      else {
        o << "  ";
      }
    }
    o << " \n";
  }
  return o;
}

void UnitTests() {
  auto Btest0 = Burrow<2>("...........", {"01", "32", "21", "03"});
  auto Btest0_next = Btest0.MoveToHallway(2);
  // for (auto const& B : Btest0_next) {
  //   std::cout << B;
  // }
  auto Btest1 = Btest0_next[4];
  if (Btest1.energy_usage_ != 40) {
    
    std::cout << "Btest1.energy_usage_ 40 != " << Btest1.energy_usage_;
    exit(1);
  }
  if (!Btest1.MoveToHallway(2).empty()) {
    std::cout << "Error";
    exit(1);
  }
  auto Btest1_next = Btest1.MoveToHallway(1);
  // for (auto const& B : Btest1_next) {
  //   std::cout << B;
  // }
  auto Btest2 = Btest1_next[0].MoveToBurrow(5);
  if (!Btest2) {
    std::cout << "Error";
    exit(1);
  }
  if (Btest2->energy_usage_ != 440) {
    
    std::cout << "Btest2.energy_usage_ 440 != " << Btest2->energy_usage_;
    exit(1);
  }
  if (!Btest2->MoveToHallway(2).empty()) {
    std::cout << "Error";
    exit(1);
  }
}

template <int T>
struct BurrowHash {
  constexpr std::size_t operator()(Burrow<T> const& burrow) const {
    std::size_t result = 0;
    for (int n = 0; n < static_cast<int>(burrow.hallway_.size()); ++n) {
      result += n*burrow.hallway_[n];
    }
    for (int b = 0; b < kNrBurrows; ++b) {
      result *= std::hash<std::string>()(burrow.burrows_[b]);
    }
    return result;
  }
};

template <int T>
struct BurrowEqual {
  constexpr bool operator()(Burrow<T> const& lhs, Burrow<T> const& rhs) const 
  {
    return lhs.hallway_ == rhs.hallway_ && lhs.burrows_ == rhs.burrows_;
  }
};

int main()
{
  UnitTests();

  // Testcase. Costs 12521 energy
  // auto B_input = Burrow<2>("...........", {"01", "32", "21", "03"});

// Part 2 input. optimal < 55138
// ...........
//   C A B D
//   D C B A 
//   D B A C 
//   C A D B
  auto B_input = Burrow<4>("...........", {"2332", "0120", "3011", "1203"});
  std::vector<Burrow<4>> frontier = {B_input};
  std::unordered_set<
    Burrow<4>,
    BurrowHash<4>,
    BurrowEqual<4>
    > visited;
  long long optimal = std::numeric_limits<long long>::max();
  long long n = 0;
  while (!frontier.empty()) {
    ++n;

    // Consider the next one
    auto best_candidate = std::min_element(frontier.begin(), frontier.end(), [](auto const& lhs, auto const& rhs){
      return lhs.energy_usage_ < rhs.energy_usage_;
    });
    auto [next_burrow, is_new] = visited.insert(*best_candidate);
    frontier.erase(best_candidate);
    if (!is_new) {
      continue;
    }

    if (n % 1000 == 0) {
      std::cout << "\nOptimum: " << optimal << " Checked: " << visited.size() << " Work: " << frontier.size() << " Up next: \n";
      std::cout << *next_burrow;
    }

    // Propagate into new states
    auto next_states = next_burrow->Possibilities();
    for (auto const& s : next_states) {
      if (s.energy_usage_ >= optimal) {
        continue;
      } else if (s.IsDone()) {
        if (s.energy_usage_ < optimal) {
          optimal = s.energy_usage_;
          // std::cout << "Lowest energy so far: " << optimal << " checked: " << visited.size() << "\n";
        }
      } else if(auto it = visited.find(s); it != visited.end()) {
        continue;
      } else {
        frontier.push_back(s);
      }
    }
  }

  std::cout << "Finished! The optimal solution expends " << optimal << " energy.\n";
}